runtime/src/iree/base/internal/cpu.c

```diff
@@ -10,9 +10,144 @@
 
 #include "iree/base/target_platform.h"
+#include "iree/base/tracing.h"
 
 //===----------------------------------------------------------------------===//
-// iree_cpu_*
+// Platform-specific processor data queries
+//===----------------------------------------------------------------------===//
+
+#if defined(IREE_PLATFORM_ANDROID) || defined(IREE_PLATFORM_LINUX)
+
+// NOTE: not all kernel versions have all of the cap bits we need defined so as
+// a practice we always define the feature bits we need locally.
+#include <sys/auxv.h>
+
+// OR's |field_bit| into |field_value| if |hwcap_bit| is set in |hwcap_value|.
+#define IREE_SET_IF_HWCAP(hwcap_value, hwcap_bit, field_value, field_bit)   \
+  (field_value) = (field_value) | iree_all_bits_set(hwcap_value, hwcap_bit) \
+                      ? field_bit                                           \
+                      : 0
+
+#if defined(IREE_ARCH_ARM_64)
+// https://docs.kernel.org/arm64/elf_hwcaps.html
+#define IREE_HWCAP_ASIMD (1 << 1)
+#define IREE_HWCAP2_I8MM (1 << 13)
+static void iree_cpu_query_data_arch_hwcaps(uint32_t hwcap, uint32_t hwcap2,
+                                            uint64_t* out_fields) {
+  IREE_SET_IF_HWCAP(hwcap, IREE_HWCAP_ASIMD, out_fields[0],
+                    IREE_CPU_DATA_FIELD_0_AARCH64_HAVE_ASIMDDP);
+  IREE_SET_IF_HWCAP(hwcap2, IREE_HWCAP2_I8MM, out_fields[0],
+                    IREE_CPU_DATA_FIELD_0_AARCH64_HAVE_I8MM);
+}
+#else
+static void iree_cpu_query_data_arch_hwcaps(uint32_t hwcap, uint32_t hwcap2,
+                                            uint64_t* out_fields) {
+  // Not supported on this arch.
+}
+#endif  // IREE_ARCH_*
+
+#undef IREE_SET_IF_HWCAP
+
+static void iree_cpu_initialize_from_platform(iree_allocator_t temp_allocator,
+                                              uint64_t* out_fields) {
+  uint32_t hwcap = getauxval(AT_HWCAP);
+  uint32_t hwcap2 = getauxval(AT_HWCAP2);
+  iree_cpu_query_data_arch_hwcaps(hwcap, hwcap2, out_fields);
+}
+
+#else
+
+static void iree_cpu_initialize_from_platform(iree_allocator_t temp_allocator,
+                                              uint64_t* out_fields) {
+  // No implementation available. CPU data will be all zeros.
+}
+
+#endif  // IREE_PLATFORM_*
+
+//===----------------------------------------------------------------------===//
+// Architecture-specific string lookup
+//===----------------------------------------------------------------------===//
+
+#define IREE_TEST_FIELD_BIT(field_key, field_value, bit_value)          \
+  if (iree_string_view_equal(key, IREE_SV(field_key))) {                \
+    *out_value = iree_all_bits_set((field_value), (bit_value)) ? 1 : 0; \
+    return true;                                                        \
+  }
+
+#if defined(IREE_ARCH_ARM_64)
+
+static bool iree_cpu_lookup_data_by_key_for_arch(
+    const uint64_t* fields, iree_string_view_t key,
+    int64_t* IREE_RESTRICT out_value) {
+  IREE_TEST_FIELD_BIT("asimddp", fields[0],
```

Comment:
There are multiple sets of strings in existence to refer to these features. Can we stick to LLVM (equivalently, GCC) cpu-features strings that are what we need to pass to the compiler? The string for dot-product instructions in "dotprod".   For i8mm, it stays "i8mm".

Reply:
Sure - if you have a table that can be referenced and how they map that'd be useful.

Reply:
For aarch64 it's in this file: https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/AArch64/AArch64.td#L295

Reply:
That link doesn't give the mapping with the other set of strings, though. The best I'm able to find is https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/AArch64/AArch64InstrInfo.td giving the mapping between LLVM feature strings and instruction mnemonics.

The other set of strings like `asimddp` it's not even 100% clear who owns it. These strings are somewhat but not perfectly in agreement with the ARM reference manual. For example, in the ARM reference manual it's `DP` not `asimddp`.  I actually believe that the strings like `asimddp` are Linux specific.  The complete mapping between the two sets of strings might be as complex as this 4-deep chain:

```
LLVM strings like "dotprod"
^
| https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/AArch64/AArch64InstrInfo.td
V
Instruction mnemonics like "sdot"
^
| ARMv8-a architecture reference manual (online PDF from ARM.com)
| Description of each instruction like "sdot"
V
ARM feature identification mnemonics like "DP"
^
| ARMv8-a architecture reference manual (online PDF from ARM.com)
| Description of control registers bitfield entries like "DP"
V
ARM feature identification bit values
^
| Linux kernel implementation of /proc/cpuinfo and `HWCAP` - usually in same patch set.
V
Linux strings like "asimddp"
```

Reply:
I really like your suggestion of following LLVM here - will make it really easy to track across cmake, the compiler/codegen, user command lines for iree-compile, and the runtime!